llvm/tools/llvm-exegesis/lib/RISCV/RISCVCounters.cpp

```diff
@@ -0,0 +1,100 @@
+//===-- RISCVCounters.cpp ---------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines RISC-V perf counters.
+//
+//===----------------------------------------------------------------------===//
+
+#include "RISCVCounters.h"
+
+namespace llvm {
+namespace exegesis {
+
+// This implementation of RISCV target for Exegesis doesn't use libpfm
+// and provides manual implementation of performance counters.
+
+inline uint64_t getRISCVCpuCyclesCount() {
+#if defined(__riscv) && defined(__linux__)
+#if __riscv_xlen == 32
+  uint32_t cycles_lo, cycles_hi0, cycles_hi1;
+  asm volatile("rdcycleh %0\n"
+               "rdcycle %1\n"
+               "rdcycleh %2\n"
+               "sub %0, %0, %2\n"
+               "seqz %0, %0\n"
+               "sub %0, zero, %0\n"
+               "and %1, %1, %0\n"
+               : "=r"(cycles_hi0), "=r"(cycles_lo), "=r"(cycles_hi1));
+  return static_cast<uint64_t>((static_cast<uint64_t>(cycles_hi1) << 32) |
+                               cycles_lo);
+#else
+  uint64_t cycles;
+  asm volatile("rdcycle %0" : "=r"(cycles));
+  return static_cast<uint64_t>(cycles);
+#endif
+#else
+  return 0;
+#endif
+}
+
+class RISCVCpuCyclesCounter : public pfm::CounterGroup {
+  uint64_t StartValue;
+  uint64_t EndValue;
+  uint64_t MeasurementCycles;
+
+public:
+  explicit RISCVCpuCyclesCounter(pfm::PerfEvent &&Event);
+
+  void start() override { StartValue = getRISCVCpuCyclesCount(); }
+
+  void stop() override { EndValue = getRISCVCpuCyclesCount(); }
+
+  Expected<llvm::SmallVector<int64_t, 4>>
+  readOrError(StringRef FunctionBytes) const override;
+};
+
+RISCVCpuCyclesCounter::RISCVCpuCyclesCounter(pfm::PerfEvent &&Event)
+    : CounterGroup(std::move(Event), {}) {
+#if defined(__riscv) && defined(__linux__)
+  char Value[2] = "0";
+  int File = 0;
+  std::error_code FileError =
+      sys::fs::openFileForRead("/proc/sys/kernel/perf_user_access", File);
```

Comment:
This is not going to be available in the future. Access to rdcycle, and allowing the kernel to configure userspace access, is deprecated, and will be removed.

Try something with rdtime instead of rdcycle

Reply:
> This is not going to be available in the future. Access to rdcycle, and allowing the kernel to configure userspace access, is deprecated, and will be removed.

Emm, we have to use libpfm to access cycle :(

See https://docs.kernel.org/admin-guide/sysctl/kernel.html#perf-user-access-arm64-and-riscv-only.

> When set to 0, user space access is disabled.
> The default value is 1, user space can read performance monitor counter registers through perf, any direct access without perf intervention will trigger an illegal instruction.
> When set to 2, which enables legacy mode (user space has direct access to cycle and insret CSRs only). Note that this legacy value is deprecated and will be removed once all user space applications are fixed.
> Note that the time CSR is always directly accessible to all modes.

> Try something with rdtime instead of rdcycle

The result may be affected by dynamic frequency scaling.

Reply:
How many RISC-V boards have a cycles counter setup in `perf`? libpfm has a counter name (that I forgot it's `PERF*CYCLES*` or something) that allows you to just use whatever perf has setup for cycles.

Reply:
In libpfm we have not the support for RISC-V. And this code can't run now because we don't support running in JIT mode. This part of the code is rather a start for future contributions.

Reply:
I wasn't asking if libpfm had support, I was asking if there was support in the perf subsystem. You can tell libpfm to use the counters setup in perf.

Also, if this doesn't support running in JIT mode, then why is all the infrastructure to run actual benchmarking being added in this patch? It would be better to split it.

Reply:
I'm pretty sure the JIT works out of the box for RISC-V. Perhaps you want to set `LLVM_TARGET_ARCH` to riscv64? And also set `LLVM_HOST_TRIPLE` to riscv64-linux-gnu for good measure.

Reply:
```
$ ./llvm-exegesis -mode=latency -mtriple=riscv64-linux-gnu -opcode-name=ADD --benchmark-phase=measure --benchmarks-file=benchmarks.yaml
llvm-exegesis error: benchmarking unavailable, LLVM was built without libpfm. You can pass --benchmark-phase=... to skip the actual benchmarking or --use-dummy-perf-counters to not query the kernel for real event counts.
$ ./llvm-exegesis -mode=latency -mtriple=riscv64-linux-gnu -opcode-name=ADD --benchmark-phase=measure --benchmarks-file=benchmarks.yaml --use-dummy-perf-counters
llvm-exegesis error: Unsupported performance counter 'not-really-an-event'
```
Llvm-exegesis doesn't run in jit mode. Lets leave jit functionality because it dosen't break build and leave it for future updates.